wire-ios-data-model/Source/MLS/Migration/ProteusToMLSMigrationCoordinator.swift

```diff
@@ -150,42 +186,52 @@ public class ProteusToMLSMigrationCoordinator: ProteusToMLSMigrationCoordinating
     /// This method is responsible for migrating all `mixed` group conversations to `mls`.
     /// It evaluates each conversation to determine if the migration needs to be finalized (i.e: updating the protocol from `mixed` to `mls`)
     /// or if it should first join the corresponding MLS group.
-    func migrateOrJoinGroupConversations() async {
+    func migrateOrJoinGroupConversations() async throws {
 
-        let mlsGroupIds = await context.perform { [self] in
-            do {
-                let conversations = try ZMConversation.fetchAllTeamGroupConversations(
-                    messageProtocol: .mixed,
-                    in: context
-                )
-                return conversations.compactMap { $0.mlsGroupID }
-            } catch {
-                logger.warn("Can't fetch conversations with `mixed` protocol")
-                return [MLSGroupID]()
-            }
-        }
+        let tuples = try await fetchMixedConversations()
 
         let mlsService = await context.perform { self.context.mlsService }
 
         guard let mlsService else {
             return logger.warn("can't migrate conversations to mls: missing `mlsService`")
         }
 
-        for groupID in mlsGroupIds {
+        for tuple in tuples {
             do {
-                if mlsService.conversationExists(groupID: groupID) {
-                    // TODO: if conversation exists we finalize migration
+                if mlsService.conversationExists(groupID: tuple.groupID) {
+                    await finalizeConversationMigration(conversation: tuple.conversation)
                 } else {
-                    try await mlsService.joinGroup(with: groupID)
+                    try await mlsService.joinGroup(with: tuple.groupID)
                 }
 
             } catch {
-                logger.warn("Can't migrate conversation with group id \(groupID.safeForLoggingDescription) to mls: \(String(describing: error))")
+                logger.warn("Can't migrate conversation with group id \(tuple.groupID.safeForLoggingDescription) to mls: \(String(describing: error))")
             }
         }
 
     }
 
+    typealias GroupIDConversationTuple = (groupID: MLSGroupID, conversation: ZMConversation)
+
+    private func fetchMixedConversations() async throws -> [GroupIDConversationTuple] {
+        return try await context.perform { [self] in
+
+            let conversations = try ZMConversation.fetchAllTeamGroupConversations(
+                messageProtocol: .mixed,
+                in: context
+            )
+
+            let tuples: [(MLSGroupID, ZMConversation)] = conversations.compactMap {
+                guard let groupID = $0.mlsGroupID else {
+                    return nil
+                }
+                return (groupID: groupID, conversation: $0)
+            }
+
+            return tuples
```

Comment:
Why do you want to return tuple? `groupID` is a `conversation.mlsGroupID`

Reply:
Yes, but it can only be accessed from the context queue. If we don't return the `groupID` here, we'll have to switch to the context queue many times when looping through the list of conversations.

Reply:
What @David-Henner said! 👍  We already discussed it before implementing.